Consistent failures with rpc_test

Since PR #1315, CI has been extremely brittle. #1315 introduced 0-hop and 2-hop virtual funding operations, mediated by the golang rpc_client, to the test suite (and CI). Previously, only the 1-hop virtual fund was under test.

recent example failures:

https://github.com/statechannels/go-nitro/actions/runs/5177367315/jobs/9327375026#step:7:578
```
...
rpc_test.go:423: bobVirtual received {0x981e13c023e2da0069ed544cc82c4ac7db43d7d1931eda38cca49bc127a9ef2c Complete {0x0000000000000000000000000000000000000000 0x6813Eb9362372EEF6200f3b1dbC3f819671cBA69 0x7E5F4552091A69125d5DfCb7b8C2659029395Bdf 0x1 0x63}}+
panic: send on closed channel

goroutine 2649 [running]:
github.com/statechannels/go-nitro/client/notifier.(*ledgerChannelListeners).Notify(0xc000258d20, {{0xe3, 0xb1, 0xaf, 0x5b, 0xd3, 0x40, 0x15, 0xf9, 0x1f, ...}, ...})
	/home/runner/work/go-nitro/go-nitro/client/notifier/listeners.go:89 +0x173
github.com/statechannels/go-nitro/client/notifier.(*ChannelNotifier).NotifyLedgerUpdated(0xc000a37770, {{0xe3, 0xb1, 0xaf, 0x5b, 0xd3, 0x40, 0x15, 0xf9, 0x1f, ...}, ...})
	/home/runner/work/go-nitro/go-nitro/client/notifier/channel-notifier.go:61 +0x1f8
github.com/statechannels/go-nitro/client.(*Client).handleEngineEvents(0xc0004c7860)
	/home/runner/work/go-nitro/go-nitro/client/client.go:107 +0x418
created by github.com/statechannels/go-nitro/client.New
	/home/runner/work/go-nitro/go-nitro/client/client.go:76 +0xbd3
FAIL	github.com/statechannels/go-nitro/client_test	0.935s
``` 
Here, sending on a closed channel generally indicates a go-nitro client attempting to communicate a protocol update message *after* it has externally been triggered to shut down (the closed channel in question is part of the event stream between go-nitro rpc-server and rpc-clients).

https://github.com/statechannels/go-nitro/actions/runs/5159120522/jobs/9293632835#step:7:751
```
...
     rpc_test.go:428: bobVirtual received unexpected notification: {0x81fc42e3c4f670506246ede8ed0cc8ba2cac129ae2f76ec04bd2eb9b3172ba81 Complete {0x0000000000000000000000000000000000000000 0x6813Eb9362372EEF6200f3b1dbC3f819671cBA69 0x7E5F4552091A69125d5DfCb7b8C2659029395Bdf 0x0 0x64}}
--- FAIL: TestRpcWithWebsockets (0.24s)
```
Here, failing with an unexpected notification - indication that a (maybe severe?) message reordering has occurred. Should be a rare flicker, but happening frequently since #1315.


https://github.com/statechannels/go-nitro/pull/1318 is a stub PR that retains the refactors to `rpc_test.go` present in #1315, but removes the 0-hop and 2-hop test runs. It has a better, but still not great, success rate.

--- a/channel/virtual.go
+++ b/channel/virtual.go
@@ -39,3 +39,10 @@ func (v *VirtualChannel) Clone() *VirtualChannel {
 	w := VirtualChannel{*v.Channel.Clone()}
 	return &w
 }
+
+func (v *VirtualChannel) GetPaidAndRemaining() {
+	remaining := v.SignedStateForTurnNum[v.latestSupportedStateTurnNum].State().Outcome[0].Allocations[0].Amount
+	paid := v.SignedStateForTurnNum[v.latestSupportedStateTurnNum].State().Outcome[0].Allocations[1].Amount
+
+	return paid,remaining
+}

--- a/channel/virtual.go
+++ b/channel/virtual.go
@@ -2,6 +2,7 @@ package channel
 
 import (
 	"errors"
+	"math/big"
 
 	"github.com/statechannels/go-nitro/channel/state"
 )
@@ -40,9 +41,9 @@ func (v *VirtualChannel) Clone() *VirtualChannel {
 	return &w
 }
 
-func (v *VirtualChannel) GetPaidAndRemaining() {
+func (v *VirtualChannel) GetPaidAndRemaining() (*big.Int, *big.Int) {
 	remaining := v.SignedStateForTurnNum[v.latestSupportedStateTurnNum].State().Outcome[0].Allocations[0].Amount
 	paid := v.SignedStateForTurnNum[v.latestSupportedStateTurnNum].State().Outcome[0].Allocations[1].Amount
 
-	return paid,remaining
+	return paid, remaining
 }

--- a/client/engine/engine.go
+++ b/client/engine/engine.go
@@ -635,10 +635,11 @@ func (e *Engine) generateNotifications(o protocols.Objective) (EngineEvent, erro
 		switch c := rel.(type) {
 		case *channel.VirtualChannel:
 			var paid, remaining *big.Int
-		
+
 			if !c.FinalCompleted() {
-				// If the channel is open, we inspect vouchers for that channel to get the future resolvable balance 
-				paid, remaining, err := query.GetVoucherBalance(c.Id, e.vm)
+				// If the channel is open, we inspect vouchers for that channel to get the future resolvable balance
+				var err error
+				paid, remaining, err = query.GetVoucherBalance(c.Id, e.vm)
 				if err != nil {
 					return outgoing, err
 				}

--- a/client/engine/engine.go
+++ b/client/engine/engine.go
@@ -649,7 +649,6 @@ func (e *Engine) generateNotifications(o protocols.Objective) (EngineEvent, erro
 				paid, remaining = c.GetPaidAndRemaining()
 			}
 			info, err := query.ConstructPaymentInfo(&c.Channel, paid, remaining)
-			// If Bob's payment channel was updated by Alice's FINAL state, then paid/remaining could be out of date here
 			if err != nil {
 				return outgoing, err
 			}

--- a/client/engine/engine.go
+++ b/client/engine/engine.go
@@ -389,15 +389,20 @@ func (e *Engine) handleMessage(message protocols.Message) (EngineEvent, error) {
 		if !ok {
 			return EngineEvent{}, fmt.Errorf("could not fetch channel for voucher %+v", voucher)
 		}
-		paid, remaining, err := query.GetVoucherBalance(c.Id, e.vm)
-		if err != nil {
-			return EngineEvent{}, err
+
+		// Vouchers only count as payment channel updates if the channel is open.
+		if !c.FinalCompleted() {
+
+			paid, remaining, err := query.GetVoucherBalance(c.Id, e.vm)
+			if err != nil {
+				return EngineEvent{}, err
+			}
+			info, err := query.ConstructPaymentInfo(c, paid, remaining)
+			if err != nil {
+				return EngineEvent{}, err
+			}
+			allCompleted.PaymentChannelUpdates = append(allCompleted.PaymentChannelUpdates, info)
 		}
-		info, err := query.ConstructPaymentInfo(c, paid, remaining)
-		if err != nil {
-			return EngineEvent{}, err
-		}
-		allCompleted.PaymentChannelUpdates = append(allCompleted.PaymentChannelUpdates, info)
 
 	}
 	return allCompleted, nil
@@ -629,11 +634,22 @@ func (e *Engine) generateNotifications(o protocols.Objective) (EngineEvent, erro
 	for _, rel := range o.Related() {
 		switch c := rel.(type) {
 		case *channel.VirtualChannel:
-			paid, remaining, err := query.GetVoucherBalance(c.Id, e.vm)
-			if err != nil {
-				return outgoing, err
+			var paid, remaining *big.Int
+		
+			if !c.FinalCompleted() {
+				// If the channel is open, we inspect vouchers for that channel to get the future resolvable balance 
+				paid, remaining, err := query.GetVoucherBalance(c.Id, e.vm)
+				if err != nil {
+					return outgoing, err
+				}
+			} else {
+				// If the channel is closed, vouchers have already been resolved.
+				// Note that when virtual defunding, this information may in fact be more up to date than
+				// the voucher balance due to a race condition https://github.com/statechannels/go-nitro/issues/1323
+				paid, remaining = c.GetPaidAndRemaining()
 			}
 			info, err := query.ConstructPaymentInfo(&c.Channel, paid, remaining)
+			// If Bob's payment channel was updated by Alice's FINAL state, then paid/remaining could be out of date here
 			if err != nil {
 				return outgoing, err
 			}

